### PR TITLE
Fixed compression middleware no gzipping client contents and added debounce + history to client location hash on Playground website

### PR DESCRIPTION
**Pull request by joshterrill**
This pull request contains two things.

First, on the client server, the `compression()` middleware was defined below the express static middleware definitions which means none of the static files were being sent gzipped. This can be shown in these screenshots which shows the file size and the response headers (notice how big the file sizes are and there is no gzip in the response header):

Compared to these screenshots which show the significantly reduced file size as well as the gzip header being present:

Another thing that I added was the ability for a user to use their back button as a history, and when you press back, it actually reverts your code back to what it was at that time. To make this not so cumbersome, I added a debounce every 1sec to write to the URL hash instead of it writing to the URL hash every key stroke.

Please let me know if there's anything I should adjust.

### CODE REVIEW

File: website/playground/static/index.html [hunk 1]
```diff
@@ -128,12 +129,15 @@
...
+watch.debounceTime(1000).subscribe(e => { 
   location.hash = btoa(codeEditor.getValue());
 });
 
-if (location.hash) {
+window.onhashchange = function(event) {
+  var cursorPosition = codeEditor.getCursor();
   codeEditor.setValue(atob(location.hash.slice(1)));
+  codeEditor.setCursor(cursorPosition);
 }
```
Comment (josephsavona) [inline website/playground/static/index.html]: does this fire on page load? we want to make sure that if you navigate to a playground URL that the correct code populates.

File: website/playground/static/index.html [hunk 2]
```diff
@@ -22,6 +22,7 @@
 <script src="third-party/lint.js"></script>
 <script src="third-party/hardwrap.js"></script>
 <script src="third-party/sublime.js"></script>
+<script src="third-party/rx.min.js"></script>
```
Comment (josephsavona) [inline website/playground/static/index.html]: Rx is a huge library, this bloats the page unnecessarily. How about either using something like lodash's per-method packages or just implementing `debounce` locally?

File: website/playground/static/index.html (hunk 1 above)
Comment (joshterrill) [reply website/playground/static/index.html]: Yes this works on load.

File: website/playground/static/index.html (hunk 2 above)
Comment (joshterrill) [reply website/playground/static/index.html]: Yeah I can do something like that.

**Comment by joshterrill** [comment]
@josephsavona I implemented a custom debounce that is only about a dozen lines long. There also was an issue with the hash not setting on the code editor like you asked, I was mistaken. That has been fixed. It works as expected now! Let me know if there's anything else you'd like.

File: website/playground/static/index.html [hunk 3]
```diff
@@ -128,12 +129,18 @@
   search: document.getElementById('search'),
 }, {lineNumbers: true});
 
-codeEditor.on('change', function() {
+codeEditor.setValue(atob(location.hash.slice(1)));
```
Comment (josephsavona) [inline website/playground/static/index.html]: should we keep the `if (location.hash) { ... }` around this?

Comment (joshterrill) [reply website/playground/static/index.html]: Sure. Done.

**Comment by joshterrill** [comment]
@josephsavona I wrapped that `setValue` in a check for `location.hash` like you asked.

**Comment by vjeux** [comment]
Thanks!

**Pull request merged**